
Sanity check for ComBat in the SVA tutorial.
========================================================

```{r, results='hide', echo=FALSE}
starttime = Sys.time()
```
`r as.character(starttime)`

## Overview
This small report aim to show that the application of ComBat as described in the [user guide](http://www.bioconductor.org/packages/2.13/bioc/vignettes/sva/inst/doc/sva.pdf) accompanying the R package [sva](http://www.bioconductor.org/packages/2.13/bioc/html/sva.html) presented in [Leek et al.](http://bioinformatics.oxfordjournals.org/cgi/pmidlookup?view=long&pmid=22257669) will lead to false results even for random data.

One sanity check will be performed

- Real data will be replaced with random numbers and a batch effect will be added wherafter the rest of the tutorial will be performed with a special focus on the procuced p-values.

Most code are taken directly from the vignette source 'vignettes/sva/inst/doc/sva.Rnw'. For more explanation, please read the [sva tutorial](http://www.bioconductor.org/packages/2.13/bioc/vignettes/sva/inst/doc/sva.pdf) in parallel while remembering that the real data is replaced with random numbers.

```{r}

options(width=65)
library(sva)
library(bladderbatch)
data(bladderdata)
library(pamr)
library(limma)

pheno = pData(bladderEset)
edata = exprs(bladderEset)

```

In this sanity check the real data is replaced with random numbers. A batch effect is also added since sva needs a batch effect or it will report "No significant surrogate variables". This is a simple and not necessarily the most realistic way of adding a batch effect.
```{r tidy=FALSE}
set.seed(100)
edata[,] = rnorm(length(edata))
for(batch in (unique(pheno$batch)))
{
	for(s in 1:dim(edata)[1])
	{
		thisgenesbatcheffect = rnorm(1)
		a= pheno$batch==batch
		edata[s, a] = edata[s, a] + thisgenesbatcheffect
	}
}

```

The first example shows the p-values when batch is ignored.
```{r dev='svg', fig.width=6, fig.height=6}

mod = model.matrix(~as.factor(cancer), data=pheno)

mod0 = model.matrix(~1,data=pheno)

n.sv = num.sv(edata,mod,method="leek")
n.sv

svobj = sva(edata,mod,mod0,n.sv=n.sv)

pValues = f.pvalue(edata,mod,mod0)
qValues = p.adjust(pValues,method="BH")
hist(pValues,breaks=100)

```

The above plot is from p-values calculated when batch is ignored, i.e. the included batch effect acts as a confounder.


```{r dev='svg', fig.width=6, fig.height=6}

modSv = cbind(mod,svobj$sv)
mod0Sv = cbind(mod0,svobj$sv)

pValuesSv = f.pvalue(edata,modSv,mod0Sv)
qValuesSv = p.adjust(pValuesSv,method="BH")
hist(pValuesSv,breaks=100)
```

This is the p-values when batch is handled by the sva method. The plot makes sense for random numbers.

```{r dev='svg', fig.width=6, fig.height=6}

fit = lmFit(edata,modSv)

contrast.matrix <- cbind("C1"=c(-1,1,0,rep(0,svobj$n.sv)),"C2"=c(0,-1,1,rep(0,svobj$n.sv)),"C3"=c(-1,0,1,rep(0,svobj$n.sv)))
fitContrasts = contrasts.fit(fit,contrast.matrix)

eb = eBayes(fitContrasts)
topTableF(eb, adjust="BH")
hist(eb$p.value,breaks=100)

```

This is the p-values when limma is used in conjunction with the sva-method. There is a slight skewness, but not much.



Next, the ComBat method is applied, where the batch effect is estimated while trying to retain the group differences. The data is adjusted and in the next step a differential expression analysis is performed.
```{r dev='svg', fig.width=6, fig.height=6}

batch = pheno$batch

mod = model.matrix(~as.factor(cancer), data=pheno)

combat_edata = ComBat(dat=edata, batch=batch, mod=mod, numCovs=NULL, par.prior=TRUE, prior.plots=FALSE)

pValuesComBat = f.pvalue(combat_edata,mod,mod0)
qValuesComBat = p.adjust(pValuesComBat,method="BH")
hist(pValuesComBat,breaks=100)

```

The p-values from the analysis when ComBat adjusted data is used shows a rather huge skewness, and many genes would falsely have been reported as differentially expressed.

The next example is in the tutroial described as a simplified version of ComBat.
```{r dev='svg', fig.width=6, fig.height=6}

modBatch = model.matrix(~as.factor(cancer) + as.factor(batch),data=pheno)
mod0Batch = model.matrix(~as.factor(batch),data=pheno)
pValuesBatch = f.pvalue(edata,modBatch,mod0Batch)
qValuesBatch = p.adjust(pValuesBatch,method="BH")
hist(pValuesBatch,breaks=100)
```

The above method of handling batch is quite similar to the blocking method described in the [limma guide](http://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf). The resulting p-values are as expected for random data.


The last example from the sva tutorial is how to use sva in a prediction analysis. This might not be relevant, but is included anyways.
```{r}

n.sv = num.sv(edata,mod,vfilter=2000,method="leek")
svobj = sva(edata,mod,mod0,n.sv=n.sv,vfilter=2000)

set.seed(12354)
trainIndicator = sample(1:57,size=30,replace=F)
testIndicator = (1:57)[-trainIndicator]

trainData = edata[,trainIndicator]
testData = edata[,testIndicator]

trainPheno = pheno[trainIndicator,]
testPheno = pheno[testIndicator,]

mydata = list(x=trainData,y=trainPheno$cancer)
mytrain = pamr.train(mydata)
table(pamr.predict(mytrain,testData,threshold=2),testPheno$cancer)
```
```{r}

trainMod = model.matrix(~cancer,data=trainPheno)
trainMod0 = model.matrix(~1,data=trainPheno)
trainSv = sva(trainData,trainMod,trainMod0)

fsvaobj = fsva(trainData,trainMod,trainSv,testData)
mydataSv = list(x=fsvaobj$db,y=trainPheno$cancer)
mytrainSv = pamr.train(mydataSv)
table(pamr.predict(mytrainSv,fsvaobj$new,threshold=1),testPheno$cancer)

```

The prediction does not improve when sva is used to adjust for batch. This makes sense since it is random data.




### References


Johnson, WE, Rabinovic, A, and Li, C (2007). Adjusting batch effects in microarray expression data using Empirical Bayes methods. Biostatistics 8(1):118-127.

Storey, J. D. and Tibshirani, R. (2003) Proc Natl Acad Sci U S A, 100, 9440-5.

Leek JT, Johnson WE, Parker HS, Jaffe AE, Storey JD.(2012) The sva package for removing batch effects and other unwanted variation in high-throughput experiments. Bioinformatics. 2012 Mar 15;28(6):882-3.
http://bioinformatics.oxfordjournals.org/cgi/pmidlookup?view=long&pmid=22257669

Smyth, GK (2005). Limma: linear models for microarray data. In: 'Bioinformatics and Computational Biology Solutions
  using R and Bioconductor'. R. Gentleman, V. Carey, S. Dudoit, R. Irizarry, W. Huber (eds), Springer, New York, pages
  397-420.
  
  R Core Team (2013). R: A language and environment for statistical computing. R Foundation for Statistical Computing,
  Vienna, Austria. URL http://www.R-project.org/

  Yihui Xie (2013). knitr: A general-purpose package for dynamic report generation in R. R package version 1.5.

  Yihui Xie (2013) Dynamic Documents with R and knitr. Chapman and Hall/CRC. ISBN 978-1482203530

  Yihui Xie (2013) knitr: A Comprehensive Tool for Reproducible Research in R. In Victoria Stodden, Friedrich Leisch and
  Roger D. Peng, editors, Implementing Reproducible Computational Research. Chapman and Hall/CRC. ISBN 978-1466561595
  
  RStudio Team (2012). RStudio: Integrated Development for R. RStudio, Inc., Boston, MA URL http://www.rstudio.com/.


```{r sessionInfo, comment=""}
sessionInfo()
```

generation ended `r as.character(Sys.time())`. Time spent `r  as.integer(round(difftime(Sys.time(),starttime, units="mins")))` minutes .
